File: languages/java/src/main/java/com/bloock/sdk/entity/identity/CredentialProof.java
package com.bloock.sdk.entity.identity;

import com.bloock.sdk.bridge.proto.BloockIdentityEntities;

/**
 * Represents the proof associated with a credential, including signature and
 * sparse merkle tree proof.
 */
public class CredentialProof {
  private final String signatureProof;
  private final String sparseMtProof;

  /**
   * Constructs an CredentialProof object with the specified parameters.
   * 
   * @param signatureProof
   * @param sparseMtProof
   */
  public CredentialProof(String signatureProof, String sparseMtProof) {
    this.signatureProof = signatureProof;
    this.sparseMtProof = sparseMtProof;
  }

  public static CredentialProof fromProto(BloockIdentityEntities.CredentialProof res) {
    return new CredentialProof(res.getSignatureProof(), res.getSparseMtProof());
  }

  public BloockIdentityEntities.CredentialProof toProto() {
    BloockIdentityEntities.CredentialProof.Builder builder = BloockIdentityEntities.CredentialProof.newBuilder();
    if (!this.signatureProof.isEmpty()) {
      builder.setSignatureProof(this.signatureProof);
    }

    if (!this.sparseMtProof.isEmpty()) {
      builder.setSparseMtProof(this.sparseMtProof);
    }

    return builder.build();
  }

  /**
   * Retrieve signature proof with string format
   * 
   * @return
   */
  public String getSignatureProof() {
    return signatureProof;
  }

  /**
   * Retrieve sparse merkle tree proof with string format
   * 
   * @return
   */
  public String getSparseMtProof() {
    return sparseMtProof;
  }
}

File: languages/js/src/entity/identity_v2/credential_proof.ts
import * as identityEntitiesProto from "../../bridge/proto/identity_entities_v2";

/**
 * Represents the proof associated with a credential, including signature and sparse merkle tree proof.
 */
export class CredentialProof {
  signatureProof: string;
  sparseMtProof?: string;

  /**
   * Constructs an CredentialProof object with the specified parameters.
   * @param signatureProof 
   * @param sparseMtProof 
   */
  constructor(signatureProof: string, sparseMtProof?: string) {
    this.signatureProof = signatureProof;
    this.sparseMtProof = sparseMtProof;
  }

  public toProto(): identityEntitiesProto.CredentialProofV2 {
    return identityEntitiesProto.CredentialProofV2.fromPartial({
      signatureProof: this.signatureProof,
      sparseMtProof: this.sparseMtProof
    });
  }

  static fromProto(
    r: identityEntitiesProto.CredentialProofV2
  ): CredentialProof {
    return new CredentialProof(r.signatureProof, r.sparseMtProof);
  }
}
